### PR TITLE
Add ArtistSeriesMoreSeries component

### PR DESCRIPTION
**Pull request by williardx**
The type of this PR is: **feature**

This PR resolves [**[FX-2015]**](https://artsyproduct.atlassian.net/browse/FX-2015)

### Description

- Adds `ArtistSeriesMoreSeries` component that shows additional artist series by the artist on the artist series view. This also involves updating the view controller to be able to navigate to a different artist series.
- Lightly refactors tests for `ArtistSeries` to remove duplicated tests and account for a new case
- Adds `.swiftpm` to `.gitignore`

**Edge cases**
- If there is no artist associated with the artist series, `ArtistSeriesMoreArtist` is not rendered.
- If there are no other artist series the child components of `ArtistSeriesMoreSeries` are not rendered.

**Followup work**
- [User does not see the current artist series when looking at a list of related artist series](https://artsyproduct.atlassian.net/browse/FX-2119)

### Test Plan

- Unit tests
- Visual QA

### Screenshots

Screen Shot 2020-07-28 at 2 56 28 PM

[FX-2015]: https://artsyproduct.atlassian.net/browse/FX-2015

### CODE REVIEW

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  const sortedSeries: MutableArtistSeriesList = series
+    .concat()
+    .sort((a: ArtistSeriesConnectionEdge, b: ArtistSeriesConnectionEdge) => {
+      return Math.sign((b?.node?.forSaleArtworksCount ?? 0) - (a?.node?.forSaleArtworksCount ?? 0))
+    })
+
+  return (
+    <>
+      <Separator />
```
Comment (williardx) [inline src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:73]: It seems wrong to have `Separator` inside the `ArtistSeriesMoreSeries` component since it's not really related to this component. But, within our top level `ArtistSeries` component there didn't seem to be a good way to check if the artist series had any related artist series without exposing that data to the `ArtistSeries` component. That didn't feel right either since it starts to add in more nuanced view logic into `ArtistSeries` which I'd rather be in `ArtistSeriesMoreSeries` itself.

File: src/lib/Scenes/ArtistSeries/ArtistSeries.tsx [hunk 2]
```diff
@@ -31,9 +35,15 @@ export const ArtistSeries: React.FC<ArtistSeriesProps> = ({ artistSeries }) => {
 export const ArtistSeriesFragmentContainer = createFragmentContainer(ArtistSeries, {
   artistSeries: graphql`
     fragment ArtistSeries_artistSeries on ArtistSeries {
+      artistIDs
+
       ...ArtistSeriesHeader_artistSeries
       ...ArtistSeriesMeta_artistSeries
       ...ArtistSeriesArtworks_artistSeries
+
+      artist: artists(size: 1) {
```
Comment (williardx) [inline src/lib/Scenes/ArtistSeries/ArtistSeries.tsx:44]: This query creates a bit of a strange situation. We now have `artist` and `artists`, both of which fetch data about the artist series's underlying artist. `artists` is used in `ArtistSeriesMeta`. Should that be resolved? It mostly creates confusion in testing since `artists` isn't exposed in other components.

File: src/lib/Scenes/ArtistSeries/ArtistSeries.tsx [hunk 3]
```diff
@@ -15,13 +16,16 @@ interface ArtistSeriesProps {
 }
 
 export const ArtistSeries: React.FC<ArtistSeriesProps> = ({ artistSeries }) => {
+  const artist = artistSeries.artist?.[0]
```
Comment (mzikherman) [inline src/lib/Scenes/ArtistSeries/ArtistSeries.tsx:19]: oo, TS safe navigation? (the `?.[0]`)

Comment (dzucconi) [reply src/lib/Scenes/ArtistSeries/ArtistSeries.tsx:19]: Yeah, I was surprised I never saw this before either

```ts
obj?.prop // Accessing object's property
obj?.[expr] // Optional chaining with expressions
arr?.[index] // Array item access with optional chaining
func?.(args) // Optional chaining with function calls
```

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx (hunk 1 above)
Comment (mzikherman) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:73]: There's no right answer, I think this is fine. Since the `Separator` is more layout-y, I agree with you that it should go 'up' a level or two in the component heirarchy. Plus, what if some place wants to use this component _without_ that separator?

But, as you point out then that component needs to essentially duplicate this fragment (which is kinda annoying), in order for it to determine the `totalCount`. That also can tend to lead to over-fetching (imagine this fragment gets tweaked, and so now instead of the queries being identical and so only one actually happens, two queries have to happen to fetch artist series, one just to get the count, and one to get a fuller list).

Especially if the designs call for this component to always render with the `Separator`, I would probably roll with it included here. It's all trade-offs!

File: src/lib/Scenes/ArtistSeries/ArtistSeries.tsx (hunk 2 above)
Comment (dzucconi) [reply src/lib/Scenes/ArtistSeries/ArtistSeries.tsx:44]: Are you saying there should be an `artist` field exposed?

I guess my question is if ArtistSeries can have multiple artists, then why are we limiting it to 1 here rather than having the UI anticipate the multiple artists?

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+          <OpaqueImageView
+            imageURL={artistSeriesItem.image?.url}
+            height={70}
+            width={70}
+            style={{ borderRadius: 2, overflow: "hidden" }}
+          />
+          <Flex ml={1} justifyContent="center" flexDirection="column">
+            <Sans size="3t">{artistSeriesItem.title}</Sans>
+            <Sans size="3" color="black60">
+              {`${artistSeriesItem.forSaleArtworksCount} available`}
```
Comment (dzucconi) [inline src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:41]: `non-blocking`

You can avoid the double string interpolation here:

```tsx
{artistSeriesItem.forSaleArtworksCount} available
```

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        <Flex justifyContent="center">
+          <ArrowRightIcon />
+        </Flex>
+      </Flex>
+    </TouchableWithoutFeedback>
+  )
+}
+
+export const ArtistSeriesMoreSeries: React.FC<ArtistSeriesMoreSeriesProps> = ({ artist }) => {
+  const navRef = useRef<any>()
```
Comment (dzucconi) [inline src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:54]: I **think** this would be `useRef<View | null>(null)` but I'm not sure what's typical here

File: src/lib/Scenes/ArtistSeries/ArtistSeries.tsx (hunk 2 above)
Comment (mzikherman) [reply src/lib/Scenes/ArtistSeries/ArtistSeries.tsx:44]: I might be misunderstanding your point, but I think this is fine. The 'other series by this artist' rail should have a fragment that's `on Artist` (vs. `on ArtistSeries`), so it can be reused in other places outside the artist series context, such as an artwork view.

So that would be why you wouldn't want to put the `artists(size: 1)` part in that component (which would then need to be `on ArtistSeries`).

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx (hunk 1 above)
Comment (dzucconi) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:73]: I'd probably extend `FlexProps` in the type for this component, then spread `{...rest}` props on the container, that way the outer context can decide if it wants, for instance, to put a border on it or to put a margin on it.

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    return SwitchBoard.presentNavigationViewController(navRef.current, `/artist-series/${slug}`)
+  }
+
+  const series = artist?.artistSeriesConnection?.edges ?? []
+
+  if (series.length === 0) {
+    return null
+  }
+
+  const sortedSeries: MutableArtistSeriesList = series
```
Comment (mzikherman) [inline src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:65]: I think we're relying on the database's sort, which is based on the 'featured' flag, and then the artworks count.

ie- all featured series sorted by count, and then non-featured series sorted

That's in https://github.com/artsy/gravity/blob/3054e781c0736d6da2207c471efd0c5682288ea8/app/models/artist_series.rb#L19 . 

Basically I don't think we need to sort the results from MP at all.

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx (hunk 5 above)
Comment (williardx) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:54]: tsc doesn't complain about it 🤔

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx (hunk 1 above)
Comment (williardx) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:73]: @mzikherman Makes sense! Agreed I should keep it as-is.

Comment (williardx) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:73]: @dzucconi Do you mean to do that for the whole `ArtistSeriesMoreSeries` component? Or just `Separator`?

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx (hunk 5 above)
Comment (dzucconi) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:54]: If you're explicitly casting something as an `any` it won't, but that generally we should try to avoid (if possible).

File: src/lib/Scenes/ArtistSeries/ArtistSeries.tsx (hunk 2 above)
Comment (williardx) [reply src/lib/Scenes/ArtistSeries/ArtistSeries.tsx:44]: > I guess my question is if ArtistSeries can have multiple artists, then why are we limiting it to 1 here rather than having the UI anticipate the multiple artists?

On the backend we've modeled it so that an artist series can have multiple artists but the feature is intended for one artist only at the moment and the designs only allow for one artist. 

@mzikherman That's a good point -- I'll leave it as-is.

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx (hunk 1 above)
Comment (dzucconi) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:73]: I'd remove the `Separator` probably — but it's a little difficult for me to tell from the PR. If it doesn't needed to be used anywhere else then whatever is simplest.

Generally any kind of outer border or outer margin should be avoided in a component that could be used elsewhere because that constrains how it's used. And so by extending Flex itself through the use of the rest props you allow the consumer of the component to dictate how it sits inside of itself.

```tsx
<ArtistSeriesMoreSeries borderTop="1px solid" borderColor="black10" mt={1} />
```

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx (hunk 5 above)
Comment (williardx) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:54]: Ah! Sorry, misread your comment. Looks like we predominantly give it a type of `any` in our codebase but that doesn't make it the right thing to do. I'll look into this more.

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx (hunk 1 above)
Comment (williardx) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:73]: @dzucconi Nice! Yeah that's a cleaner way to do this, I'll update it.

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  artistSeriesItem: NonNullable<NonNullable<ArtistSeriesConnectionEdge>["node"]>
+  handleNavigation: (slug: string) => void
+}
+
+export const ArtistSeriesMoreSeriesItem: React.FC<ArtistSeriesMoreSeriesItemProps> = ({
+  artistSeriesItem,
+  handleNavigation,
+}) => {
+  return (
+    <TouchableWithoutFeedback onPress={() => handleNavigation(artistSeriesItem.slug)}>
```
Comment (ds300) [inline src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:27]: We should add user feedback for this touchable. There's two basic types of touch interaction we use, outlined here https://www.notion.so/artsy/Motion-Guidelines-11000a5ea9764bf182cc0732f5cebb17#1df138bbe6c442e5a6c92722e13e2b47 

this should use the 'darken' interaction type, which can be done with `<TouchableWithHighlight underlayColor={color("black5")} ...`. You'll probably need to adjust the view hierarchy to allow the darkening area to stretch full-screen.

File: src/lib/Scenes/ArtistSeries/ArtistSeries.tsx [hunk 8]
```diff
@@ -15,13 +16,23 @@ interface ArtistSeriesProps {
...
           <ArtistSeriesHeaderFragmentContainer artistSeries={artistSeries} />
           <ArtistSeriesMetaFragmentContainer artistSeries={artistSeries} />
           <ArtistSeriesArtworksFragmentContainer artistSeries={artistSeries} />
+          {!!artist && (
+            <ArtistSeriesMoreSeriesFragmentContainer
+              artist={artist}
+              borderTopWidth="1px"
+              borderTopColor="black10"
+              mt={1}
+            />
```
Comment (ds300) [inline src/lib/Scenes/ArtistSeries/ArtistSeries.tsx:34]:
```suggestion
          <>
            <Separator mt={1} />
            <ArtistSeriesMoreSeriesFragmentContainer />
          </>
```

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+          <Flex ml={1} justifyContent="center" flexDirection="column">
+            <Sans size="3t">{artistSeriesItem.title}</Sans>
+            <Sans size="3" color="black60">
+              {artistSeriesItem.forSaleArtworksCount} available
+            </Sans>
+          </Flex>
+        </Flex>
+        <Flex justifyContent="center">
+          <ArrowRightIcon />
+        </Flex>
```
Comment (ds300) [inline src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:45]: This right chevron has a 20px margin but we typically reduce it to 15px since the chevron has about 5px of right padding already and it looks good when the tip of the chevron aligns with the content margin. I think you could add `mr="-5px"` here to adopt that.

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    <TouchableWithoutFeedback onPress={() => handleNavigation(artistSeriesItem.slug)}>
+      <Flex key={artistSeriesItem.internalID} flexDirection="row" justifyContent="space-between" mb={1}>
+        <Flex flexDirection="row">
+          <OpaqueImageView
+            imageURL={artistSeriesItem.image?.url}
+            height={70}
+            width={70}
+            style={{ borderRadius: 2, overflow: "hidden" }}
+          />
+          <Flex ml={1} justifyContent="center" flexDirection="column">
```
Comment (ds300) [inline src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx:36]: `column` is the default on mobile, you don't need to specify :)

File: src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx (hunk 1 above)
Comment (williardx) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx]: @ds300 Re: https://github.com/artsy/eigen/pull/3645#discussion_r463501446, this discussion gives the reasoning around why I ended up using CSS to effectively create a separator instead of using the `Separator` component itself. Curious for your thoughts if there's a better way to do this! As of right now the current implementation has the desired behavior (no `Separator` in the `ArtistSeriesMoreSeries` component and no visible separator when there are no related artist series.)

Comment (ds300) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx]: ah I see. Ok feel free to ignore my comment. In other places we duplicate the 'will this component render anything' logic in the parent if we need to do this kind of thing but I'm OK with this workaround. Can you add a comment in the parent component to make it obvious why you've done it that way though? I don't want other people to repeat that pattern blindly.

Comment (williardx) [reply src/lib/Scenes/ArtistSeries/ArtistSeriesMoreSeries.tsx]: Yep no problem!
